subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/SplitClasspathEntrySnapshotCache.java

```diff
@@ -49,11 +49,25 @@ public ClasspathEntrySnapshot get(File file, HashCode hash) {
     }
 
     @Override
-    public ClasspathEntrySnapshot get(File entry, Factory<ClasspathEntrySnapshot> factory) {
-        if (globalCacheLocations.isInsideGlobalCache(entry.getPath())) {
-            return globalCache.get(entry, factory);
+    public ClasspathEntrySnapshot get(File entry, Function<? super File, ? extends ClasspathEntrySnapshot> factory) {
+        return getCacheFor(entry).get(entry, factory);
+    }
+
+    @Override
+    public ClasspathEntrySnapshot get(File key) {
+        return getCacheFor(key).get(key);
+    }
+
+    @Override
+    public void put(File key, ClasspathEntrySnapshot value) {
+        getCacheFor(key).put(key, value);
+    }
```

Comment:
Seems like now we need to add a bunch of methods that we don't necessarily want to use. This doesn't seem like a good change overall.

Reply:
We could have a default implementation for this guy that would throw an `UnsupportedOperationException`. Alternatively, we could have a `DirectAccessCache` or something that would introduce this method.

Reply:
I really don't want to throw `UnsupportedOperationException` for these methods if somebody starts calling them. We use the methods on "most" caches and it I wouldn't want to introduce a separate interface for these methods which are very easy to implement.

Reply:
Nah, the `UOE` was not really a serious idea. I just don't want to require that any future cache also supports `put()`, so I'd rather move that method to a separate type.

Reply:
I tried removing `put` from `Cache` and only leaving it in `PersistentIndexedCache`. Though there seem to be a few usages of `CrossBuildInMemoryCache.put()` which can't be so easily removed, the most prominent instance probably this: https://github.com/gradle/gradle/blob/45e5124fc9c65c11356259a30d51d1fd81f9ef20/subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/CrossBuildInMemoryCachingScriptClassCache.java#L41-L52

So I would actually leave `put` on `Cache`, given that the implementations I added have been trivial to add.